homework_solution.md

```diff
@@ -0,0 +1,474 @@
+       WHERE original_filename LIKE '%2017 Q4%'
+       AND start_time_str IS NOT NULL
+       AND id <> 2302635;
+       ```
+
+       (While dealing with this, we found that record `2302635` has `end_time_str` of `NULLNULL`, so we'll treat it separately)
+
+       ```
+       UPDATE trips
+       SET
+          start_time = (SELECT TO_TIMESTAMP(start_time_str, 'MM/DD/YY HH24:MI')),
+          end_time_str = NULL
+       WHERE id = 2302635;
+       ```
+
+    5. Update `2018` records
+       ```
+       UPDATE trips
+       SET
+          start_time = (SELECT TO_TIMESTAMP(start_time_str, 'MM/DD/YYYY HH24:MI')),
+          end_time = (SELECT TO_TIMESTAMP(end_time_str, 'MM/DD/YYYY HH24:MI'))
+       WHERE original_filename LIKE '%2018%'
+       AND start_time_str IS NOT NULL;
+       ```
+
+3.  _**Other than the index in class, would we benefit from any other indexes on this table? Why or why not?**_
+
+    Yes. An index on `original_filename` goes a long way here.
+    It can be perceived by the observing the different `cost` before and after the index:
+
+    ```
+
+    postgres=# EXPLAIN SELECT original_filename, (ARRAY_AGG(DISTINCT start_time_str))[1]
+    FROM trips
+    GROUP BY original_filename;
+    QUERY PLAN
+
+    ---
+
+    GroupAggregate (cost=719052.22..744667.25 rows=8 width=69)
+    Group Key: original_filename
+    -> Sort (cost=719052.22..727590.53 rows=3415324 width=52)
+    Sort Key: original_filename
+    -> Seq Scan on trips (cost=0.00..114957.24 rows=3415324 width=52)
+    JIT:
+    Functions: 7
+    Options: Inlining true, Optimization true, Expressions true, Deforming true
+    (8 rows)
+
+    ```
+
+    ```
+
+    postgres=# CREATE INDEX idx_original_filename ON trips USING btree (original_filename);
+    CREATE INDEX
+
+    ```
+
+    ```
+
+    postgres=# EXPLAIN SELECT original_filename, (ARRAY_AGG(DISTINCT start_time_str))[1]
+    FROM trips
+    GROUP BY original_filename;
+    QUERY PLAN
+
+    ---
+
+    GroupAggregate (cost=0.56..343851.30 rows=8 width=69)
+    Group Key: original_filename
+    -> Index Scan using idx_original_filename on trips (cost=0.56..326774.58 rows=3415324 width=52)
+    JIT:
+    Functions: 5
+    Options: Inlining false, Optimization false, Expressions true, Deforming true
+    (6 rows)
+
+    ```
+
+# Part 3: Data-driven insights
+
+1. _**Build a mini-report that does a breakdown of number of trips by month**_
+
+   ```
+   SELECT
+      CASE  EXTRACT (MONTH FROM start_time)
+         WHEN 1 THEN '01 - JAN'
+         WHEN 2 THEN '02 - FEB'
+         WHEN 3 THEN '03 - MAR'
+         WHEN 4 THEN '04 - APR'
+         WHEN 5 THEN '05 - MAY'
+         WHEN 6 THEN '06 - JUN'
+         WHEN 7 THEN '07 - JUL'
+         WHEN 8 THEN '08 - AUG'
+         WHEN 9 THEN '09 - SEP'
+         WHEN 10 THEN '10 - OCT'
+         WHEN 11 THEN '11 - NOV'
+         WHEN 12 THEN '12 - DEC'
+      END
+      AS Month,
+      COUNT(id) as Trips
+   FROM trips
+   GROUP BY Month
+   ORDER BY 1;
```

Comment:
[minor]
What is this actually ordering by? Can you make that more obvious?

Reply:
It's ordering by the first column. I could call it by the order name. In this case, I see it as matter of preference.
Personally, I like the flexibility of being able to add an alias later on and not break the ordering - of course, this approach removes the flexibility of reordering the columns without breaking the ordering. It's always a trade off!

Reply:
good answer! Generally speaking, I prioritize clarity over future changability in code. For example, if I wanted to change the name of that first field from `Month` to something else, I'd so a search+replace for `Month` anyway, so it wouldn't actually be any harder to update it, if it had `ORDER BY Month` as I'm imagining.

You're right about it being a trade-off though.  If changeability if your priority, then this is a better way to go.  The important thing is to make sure that your code tells the reader what you want it to.  So if your intention here was that the `ORDER BY` clause says 'Order by month name', referring to it by column number would be a mistake.  But if your intention is 'Order by the first column (no matter what that happens to be)', then this is correct!